task6/script.js

```diff
@@ -0,0 +1,427 @@
+        'author'
+    ];
+
+    const optionalProperties = [
+        'likes',
+        'photoLink',
+        'hashTags'
+    ];
+
+    const allProperties = [
+        'likes',
+        'photoLink',
+        'hashTags',
+        'id',
+        'description',
+        'createdAt',
+        'author'
+    ];
+
+    function validatePost(post) {
+        if (typeof post !== 'object') {
+            return false;
+        }
+        let optionalFieldAmount = 0;
+
+        for (let index = 0; index < requiredProperties.length; ++index) {
+            let property = requiredProperties[index];
+            if (!post.hasOwnProperty(property)) {
+                return false;
+            }
+            switch (property) {
+                case 'id':
+                case 'description':
+                case 'author': {
+                    if (typeof post[property] !== 'string') {
+                        return false;
+                    }
+                    break;
+                }
+                case 'createdAt': {
+                    if (!(post[property] instanceof  Date)) {
+                        return false;
+                    }
+                    break;
+                }
+                default:
+                    break;
+            }
+        }
+
+        for (let index = 0; index < optionalProperties.length; ++index) {
+            let property = optionalProperties[index];
+            if (post.hasOwnProperty(property)) {
+                switch (property) {
+                    case 'hashTags':
+                    case 'likes':
+                    case 'photoLink': {
+                        if (!Array.isArray(post[property])
+                            && !post[property].every(function (element, index, array) {
+                                return typeof element === 'string';
+                            })) {
+                            return false;
+                        }
+                        optionalFieldAmount++;
+                        break;
+                    }
+                    default:
+                        break;
+                }
+            }
+        }
+        return requiredProperties.length + optionalFieldAmount === Object.keys(post).length;
+    }
+
+    function compareDates(a, b) {
+        return b.createdAt - a.createdAt;
+    }
+
+
+    //all properties in filter config are optional and aren't empty
+    function validateFilterConfig(filterConfig) {
+        let optionalFieldAmount = 0;
+
+        for (let index = 0; index < allProperties.length; ++index) {
+            let property = allProperties[index];
+            if (filterConfig.hasOwnProperty(property)) {
+                switch (property) {
+                    case 'hashTags':
+                    case 'likes':
+                    case 'photoLink': {
+                        if (!Array.isArray(filterConfig[property])
+                            && !filterConfig[property].every(function (element, index, array) {
```

Comment:
Remove not needed params from callback. Also, it would be nice to use arrow functions as callbacks